diff --git a/share/cmake/Caffe2/public/cuda.cmake b/share/cmake/Caffe2/public/cuda.cmake
index 229e8b7..d01db0f 100644
--- a/share/cmake/Caffe2/public/cuda.cmake
+++ b/share/cmake/Caffe2/public/cuda.cmake
@@ -134,8 +134,9 @@ endif()
 # find lbnvrtc.so
 set(CUDA_NVRTC_LIB "${CUDA_nvrtc_LIBRARY}" CACHE FILEPATH "")
 if(CUDA_NVRTC_LIB AND NOT CUDA_NVRTC_SHORTHASH)
+  find_package(Python COMPONENTS Interpreter REQUIRED)
   execute_process(
-    COMMAND Python::Interpreter -c
+    COMMAND ${Python_EXECUTABLE} -c
     "import hashlib;hash=hashlib.sha256();hash.update(open('${CUDA_NVRTC_LIB}','rb').read());print(hash.hexdigest()[:8])"
     RESULT_VARIABLE _retval
     OUTPUT_VARIABLE CUDA_NVRTC_SHORTHASH)
