diff --git a/src/ucp/proto/proto_common.c b/src/ucp/proto/proto_common.c
index ac325f77f..d301ede3e 100644
--- a/src/ucp/proto/proto_common.c
+++ b/src/ucp/proto/proto_common.c
@@ -133,6 +133,10 @@ void ucp_proto_common_get_lane_distance(const ucp_proto_init_params_t *params,

     tl_sys_dev = context->tl_rscs[rsc_index].tl_rsc.sys_device;
     status     = ucs_topo_get_distance(sys_dev, tl_sys_dev, distance);
+    if (status != UCS_OK) {
+        *distance = ucs_topo_default_distance;
+        return;
+    }
     ucs_assertv_always(status == UCS_OK, "sys_dev=%d tl_sys_dev=%d", sys_dev,
                        tl_sys_dev);
 }
diff --git a/src/ucp/rndv/proto_rndv.c b/src/ucp/rndv/proto_rndv.c
index a7e092198..e0265c864 100644
--- a/src/ucp/rndv/proto_rndv.c
+++ b/src/ucp/rndv/proto_rndv.c
@@ -82,9 +82,11 @@ ucp_proto_rndv_ctrl_get_md_map(const ucp_proto_rndv_ctrl_init_params_t *params,
     mem_sys_dev = params->super.super.select_param->sys_dev;
     ucs_for_each_bit(ep_sys_dev, *sys_dev_map) {
         status = ucs_topo_get_distance(mem_sys_dev, ep_sys_dev, sys_distance);
-        ucs_assertv_always(status == UCS_OK, "mem_info->sys_dev=%d sys_dev=%d",
-                           mem_sys_dev, ep_sys_dev);
-        ++sys_distance;
+        if (status != UCS_OK) {
+            *sys_distance = ucs_topo_default_distance;
+        } else {
+            ++sys_distance;
+        }
     }
 }

