diff --git a/config/m4/cuda.m4 b/config/m4/cuda.m4
index a13584796..751679e7a 100644
--- a/config/m4/cuda.m4
+++ b/config/m4/cuda.m4
@@ -52,22 +52,6 @@ AS_IF([test "x$cuda_checked" != "xyes"],
                [AC_CHECK_LIB([cudart], [cudaGetDeviceCount],
                              [CUDA_LIBS="$CUDA_LIBS -lcudart"], [cuda_happy="no"])])

-         # Check nvml header files
-         AS_IF([test "x$cuda_happy" = "xyes"],
-               [AC_CHECK_HEADERS([nvml.h],
-                                 [cuda_happy="yes"],
-                                 [AS_IF([test "x$with_cuda" != "xguess"],
-                                        [AC_MSG_ERROR([nvml header not found. Install appropriate cuda-nvml-devel package])])
-                                  cuda_happy="no"])])
-
-         # Check nvml library
-         AS_IF([test "x$cuda_happy" = "xyes"],
-               [AC_CHECK_LIB([nvidia-ml], [nvmlInit],
-                             [CUDA_LIBS="$CUDA_LIBS -lnvidia-ml"],
-                             [AS_IF([test "x$with_cuda" != "xguess"],
-                                    [AC_MSG_ERROR([libnvidia-ml not found. Install appropriate nvidia-driver package])])
-                              cuda_happy="no"])])
-
          LDFLAGS="$save_LDFLAGS"

          # Check for cuda static library
diff --git a/src/uct/cuda/base/cuda_iface.h b/src/uct/cuda/base/cuda_iface.h
index bf3f83c81..b05b8eeb3 100644
--- a/src/uct/cuda/base/cuda_iface.h
+++ b/src/uct/cuda/base/cuda_iface.h
@@ -12,7 +12,9 @@
 #include <ucs/async/eventfd.h>
 #include <cuda_runtime.h>
 #include <cuda.h>
+#if 0
 #include <nvml.h>
+#endif


 #define UCT_CUDA_DEV_NAME       "cuda"
@@ -76,6 +78,7 @@
     UCT_CUDA_FUNC(_func, UCS_LOG_LEVEL_ERROR)


+#if 0
 #define UCT_NVML_FUNC(_func, _log_level)                        \
     ({                                                          \
         ucs_status_t _status = UCS_OK;                          \
@@ -94,6 +97,7 @@

 #define UCT_NVML_FUNC_LOG_ERR(_func) \
     UCT_NVML_FUNC(_func, UCS_LOG_LEVEL_ERROR)
+#endif


 #define UCT_CUDADRV_FUNC(_func, _log_level)                     \
diff --git a/src/uct/cuda/cuda_ipc/cuda_ipc_iface.c b/src/uct/cuda/cuda_ipc/cuda_ipc_iface.c
index 8cb9ad273..990aaf8ac 100644
--- a/src/uct/cuda/cuda_ipc/cuda_ipc_iface.c
+++ b/src/uct/cuda/cuda_ipc/cuda_ipc_iface.c
@@ -117,6 +117,7 @@ static double uct_cuda_ipc_iface_get_bw()
     }
 }

+#if 0
 /* calls nvmlInit_v2 and nvmlShutdown which are expensive but
  * get_device_nvlinks should be outside critical path */
 static int uct_cuda_ipc_get_device_nvlinks(int ordinal)
@@ -170,6 +171,12 @@ err_sd:
 err:
     return 0;
 }
+#else
+static int uct_cuda_ipc_get_device_nvlinks(int ordinal)
+{
+    return 0;
+}
+#endif

 static size_t uct_cuda_ipc_iface_get_max_get_zcopy(uct_cuda_ipc_iface_t *iface)
 {
