--- a/public/examples/tensor_interop/python/tensor_interop.py
+++ b/public/examples/tensor_interop/python/tensor_interop.py
@@ -19,7 +19,13 @@ import os
 
 from holoscan.core import Application, Operator, OperatorSpec
 from holoscan.data_loggers import BasicConsoleLogger
-from holoscan.operators import HolovizOp, VideoStreamReplayerOp
+from holoscan.operators import (
+    FormatConverterOp,
+    HolovizOp,
+    VideoStreamRecorderOp,
+    VideoStreamReplayerOp,
+)
+from holoscan.resources import UnboundedAllocator
 
 try:
     import cupy as cp
@@ -126,6 +132,22 @@ class MyVideoProcessingApp(Application):
         self.add_flow(source, image_processing)
         self.add_flow(image_processing, visualizer, {("", "receivers")})
 
+        recorder_format_converter = FormatConverterOp(
+            self,
+            name="recorder_format_converter",
+            in_dtype="rgba8888",
+            out_dtype="rgb888",
+            pool=UnboundedAllocator(self, name="pool"),
+        )
+        recorder = VideoStreamRecorderOp(self, name="recorder", **self.kwargs("recorder"))
+
+        visualizer.add_arg(allocator=UnboundedAllocator(self, name="allocator"))
+
+        self.add_flow(
+            visualizer, recorder_format_converter, {("render_buffer_output", "source_video")}
+        )
+        self.add_flow(recorder_format_converter, recorder)
+
 
 def main(config_file):
     app = MyVideoProcessingApp()
