diff --git a/public/examples/holoviz/cpp/holoviz_geometry.cpp b/public/examples/holoviz/cpp/holoviz_geometry.cpp
index 1eb86c2ed..9be565f8a 100644
--- a/public/examples/holoviz/cpp/holoviz_geometry.cpp
+++ b/public/examples/holoviz/cpp/holoviz_geometry.cpp
@@ -26,6 +26,12 @@

 #include <gxf/std/tensor.hpp>

+#ifdef RECORD_OUTPUT
+  #include <holoscan/operators/video_stream_recorder/video_stream_recorder.hpp>
+  #include <holoscan/operators/format_converter/format_converter.hpp>
+  #include <recorder.hpp>
+#endif
+
 namespace holoscan::ops {

 /**
@@ -316,6 +322,9 @@ class HolovizGeometryApp : public holoscan::Application {
     add_flow(source, visualizer, {{"outputs", "receivers"}});
     add_flow(source, visualizer, {{"output_specs", "input_specs"}});
     add_flow(replayer, visualizer, {{"output", "receivers"}});
+
+    // Recorder to validate the video output
+    RECORDER(visualizer);
   }

  private:
