--- a/public/examples/video_replayer/cpp/video_replayer.cpp
+++ b/public/examples/video_replayer/cpp/video_replayer.cpp
@@ -22,6 +22,12 @@
 #include <holoscan/operators/video_stream_replayer/video_stream_replayer.hpp>
 #include "holoscan/data_loggers/async_console_logger/async_console_logger.hpp"
 
+#ifdef RECORD_OUTPUT
+#include <holoscan/operators/format_converter/format_converter.hpp>
+#include <holoscan/operators/video_stream_recorder/video_stream_recorder.hpp>
+#include <recorder.hpp>
+#endif
+
 class VideoReplayerApp : public holoscan::Application {
  public:
   void on_window_closed() {
@@ -65,6 +71,9 @@ class VideoReplayerApp : public holoscan::Application {
           }));
       add_flow(replayer, visualizer2, {{"output", "receivers"}});
     }
+
+    // Recorder to validate the video output
+    RECORDER(visualizer);
   }
 };
 