diff --git a/CMakeLists.txt b/CMakeLists.txt
index 03968f4..6542d14 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -84,9 +84,7 @@ cmake_dependent_option(CLI11_BUILD_EXAMPLES_JSON "Build CLI11 json example" OFF
 
 cmake_dependent_option(CLI11_SINGLE_FILE_TESTS "Duplicate all the tests for a single file build"
                        OFF "BUILD_TESTING;CLI11_SINGLE_FILE" OFF)
-
-cmake_dependent_option(CLI11_INSTALL "Install the CLI11 folder to include during install process"
-                       ON "CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME" OFF)
+option(CLI11_INSTALL "Install the CLI11 folder to include during install process" ON)
 
 cmake_dependent_option(
   CLI11_FORCE_LIBCXX "Force clang to use libc++ instead of libstdc++ (Linux only)" OFF
@@ -181,14 +179,14 @@ if(CLI11_INSTALL)
 
   # Make version available in the install
   install(FILES "${PROJECT_BINARY_DIR}/CLI11ConfigVersion.cmake"
-          DESTINATION "${CMAKE_INSTALL_DATADIR}/cmake/CLI11")
+          DESTINATION "${CMAKE_INSTALL_LIBDIR}/cmake/CLI11")
 
   # Install the export target as a file
   install(
     EXPORT CLI11Targets
     FILE CLI11Config.cmake
     NAMESPACE CLI11::
-    DESTINATION "${CMAKE_INSTALL_DATADIR}/cmake/CLI11")
+    DESTINATION "${CMAKE_INSTALL_LIBDIR}/cmake/CLI11")
 
   # Use find_package on the installed package
   export(
diff --git a/cmake/CLI11GeneratePkgConfig.cmake b/cmake/CLI11GeneratePkgConfig.cmake
--- a/cmake/CLI11GeneratePkgConfig.cmake
+++ b/cmake/CLI11GeneratePkgConfig.cmake
@@ -5,3 +5,2 @@ endif()
 endif()
 
-install(FILES "${PROJECT_BINARY_DIR}/CLI11.pc" DESTINATION "${CMAKE_INSTALL_DATADIR}/pkgconfig")
