diff --git a/CMakeLists.txt b/CMakeLists.txt
index ea80135..bfbd3b3 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -11,6 +11,7 @@ target_include_directories(${PROJECT_NAME}
                 $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}>
                 $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/${PROJECT_NAME}/>
 )
+add_library(${PROJECT_NAME}::${PROJECT_NAME} ALIAS ${PROJECT_NAME})
 
 install(TARGETS ${PROJECT_NAME} 
     EXPORT ${PROJECT_NAME}Targets
@@ -37,8 +38,6 @@ install(EXPORT
                 "${PROJECT_NAME}::"
         DESTINATION
                 ${CMAKE_INSTALL_LIBDIR}/cmake/${PROJECT_NAME}
-        COMPONENT
-                Devel
 )
 
 install(
@@ -47,8 +46,6 @@ install(
                 ${CMAKE_CURRENT_BINARY_DIR}/${PROJECT_NAME}ConfigVersion.cmake
         DESTINATION
                 ${CMAKE_INSTALL_LIBDIR}/cmake/${PROJECT_NAME}
-        COMPONENT
-                Devel
 )
 
 install(
@@ -58,7 +55,12 @@ install(
                 lightweightsemaphore.h 
                 LICENSE.md
         DESTINATION 
-                ${CMAKE_INSTALL_INCLUDEDIR}/${PROJECT_NAME}/moodycamel
+                ${CMAKE_INSTALL_INCLUDEDIR}/${PROJECT_NAME}
+)
+export(
+        TARGETS concurrentqueue
+        NAMESPACE ${PROJECT_NAME}::
+        FILE ${PROJECT_NAME}Targets.cmake
 )
 
 set(CPACK_PACKAGE_NAME ${PROJECT_NAME})
