diff --git a/CMakeLists.txt b/CMakeLists.txt
index 5e9c1c9..0d15962 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -75,21 +75,25 @@ if(NVTX3_INSTALL)
 
     # Install headers
     install(DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/include/"
-            DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}")
+            DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}"
+            COMPONENT nvtx3)
 
     # Install the config files
     install(FILES
         "${CMAKE_CURRENT_BINARY_DIR}/nvtx3-config.cmake"
         "${CMAKE_CURRENT_BINARY_DIR}/nvtx3-config-version.cmake"
         DESTINATION "${CMAKE_INSTALL_LIBDIR}/cmake/nvtx3"
+        COMPONENT nvtx3
     )
 
     # Install the targets
-    install(TARGETS nvtx3-c nvtx3-cpp EXPORT nvtx3-targets)
+    install(TARGETS nvtx3-c nvtx3-cpp EXPORT nvtx3-targets
+            COMPONENT nvtx3)
     install(EXPORT nvtx3-targets
         FILE nvtx3-targets.cmake
         NAMESPACE nvtx3::
-        DESTINATION "${CMAKE_INSTALL_LIBDIR}/cmake/nvtx3")
+        DESTINATION "${CMAKE_INSTALL_LIBDIR}/cmake/nvtx3"
+            COMPONENT nvtx3)
     export(EXPORT nvtx3-targets
         FILE nvtx3-targets.cmake
         NAMESPACE nvtx3::)
