diff --git a/CMakeLists.txt b/CMakeLists.txt
index f5a878e..1e6abaf 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -161,8 +161,8 @@ rapids_cmake_build_type(Release)
 
 rapids_cpm_init()
 
-option(BUILD_TESTS "Configure CMake to build tests" ON)
-option(BUILD_SHARED_LIBS "Build shared libraries" ON)
+option(RAPIDS_LOGGER_BUILD_TESTS "Configure CMake to build tests" OFF)
+option(RAPIDS_LOGGER_BUILD_SHARED_LIBS "Build shared libraries" ON)
 
 include(CMakeDependentOption)
 # We cannot hide all spdlog symbols if we are building a static library.
@@ -225,7 +225,7 @@ else()
 endif()
 target_link_libraries(rapids_logger PRIVATE spdlog::spdlog)
 
-if(BUILD_TESTS)
+if(RAPIDS_LOGGER_BUILD_TESTS)
   include(CTest)
   add_subdirectory(tests)
 endif()
