diff --git a/cpp/CMakeLists.txt b/cpp/CMakeLists.txt
index 1f2241d..f6885cb 100644
--- a/cpp/CMakeLists.txt
+++ b/cpp/CMakeLists.txt
@@ -69,12 +69,18 @@ rapids_find_package(
 # add third party dependencies using CPM
 rapids_cpm_init()
 
-include(${rapids-cmake-dir}/cpm/rapids_logger.cmake)
-rapids_cpm_rapids_logger(BUILD_EXPORT_SET rmm-exports INSTALL_EXPORT_SET rmm-exports)
+if(NOT RAPIDS_LOGGER_SOURCE_DIR)
+  include(${rapids-cmake-dir}/cpm/rapids_logger.cmake)
+  rapids_cpm_rapids_logger(BUILD_EXPORT_SET rmm-exports INSTALL_EXPORT_SET rmm-exports)
+endif()
 create_logger_macros(RMM "rmm::default_logger()" include/rmm)
 
-include(cmake/thirdparty/get_cccl.cmake)
-include(cmake/thirdparty/get_nvtx.cmake)
+if(NOT CCCL_SOURCE_DIR)
+  include(cmake/thirdparty/get_cccl.cmake)
+endif()
+if(NOT nvtx3_SOURCE_DIR)
+  include(cmake/thirdparty/get_nvtx.cmake)
+endif()
 
 # ##################################################################################################
 # * library targets --------------------------------------------------------------------------------
